### PR TITLE
Use RandomizedSearchCV and point to max_leaf_nodes

### LINKED ISSUE #314: adjust notebook on hyperparameters on ensemble models
Regarding this notebook: https://inria.github.io/scikit-learn-mooc/python_scripts/ensemble_hyperparameters.html

- use `RandomizedSearchCV`
- use `max_leaf_nodes`
- larger `n_estimators`
- improve conclusion of each code snippet.

### PR DESCRIPTION
**Pull request by glemaitre**
closes #314

### CODE REVIEW

File: python_scripts/ensemble_hyperparameters.py [hunk 1]
```diff
@@ -41,52 +45,57 @@
 
 # %%
 import pandas as pd
-from sklearn.model_selection import GridSearchCV
+from scipy.stats import randint
```
Comment (ogrisel) [inline python_scripts/ensemble_hyperparameters.py:48]:
```suggestion
```

File: python_scripts/ensemble_hyperparameters.py [hunk 2]
```diff
@@ -41,52 +45,57 @@
...
+from sklearn.model_selection import RandomizedSearchCV
 from sklearn.ensemble import RandomForestRegressor
 
-param_grid = {
-    "n_estimators": [10, 20, 30],
-    "max_depth": [3, 5, None],
+param_distributions = {
+    "n_estimators": randint(10, 200),
```
Comment (ogrisel) [inline python_scripts/ensemble_hyperparameters.py:53]:
```suggestion
    "n_estimators": [1, 2, 5, 10, 20, 50, 100, 200, 500],
```

File: python_scripts/ensemble_hyperparameters.py [hunk 3]
```diff
@@ -41,52 +45,57 @@
...
+    "max_leaf_nodes": randint(3, 300),
```
Comment (ogrisel) [inline python_scripts/ensemble_hyperparameters.py:54]:
```suggestion
    "max_leaf_nodes": [2, 5, 10, 20, 50, 100],
```

File: python_scripts/ensemble_hyperparameters.py [hunk 4]
```diff
@@ -41,52 +45,57 @@
...
-# typically between 3 to 8 levels. Having very weak learners at each step will
-# help reducing overfitting.
+# Let's first discuss the `max_depth` (or `max_leaf_nodes`) parameter. We saw
+# in the section on gradient-boosting that the algorithm fits the error of the
+# previous tree in the ensemble. Thus, fitting fully grown trees will be
+# detrimental. Indeed, the first tree of the ensemble would perfectly fit
+# (overfit) the data and thus no subsequent tree would be required, since there
+# would be no residuals. Therefore, the tree used in gradient-boosting should
+# have a low depth, typically between 3 to 8 levels, or few leaves ($2^3=8$ to
+# $2^8$). Having very weak learners at each step will help reducing overfitting.
```
Comment (ogrisel) [inline python_scripts/ensemble_hyperparameters.py:98]:
```suggestion
# $2^8=256$). Having very weak learners at each step will help reducing
# overfitting.
```

File: python_scripts/ensemble_hyperparameters.py [hunk 5]
```diff
@@ -103,37 +112,46 @@
...
 # %%
+from scipy.stats import loguniform
 from sklearn.ensemble import GradientBoostingRegressor
 
-param_grid = {
-    "n_estimators": [10, 30, 50],
-    "max_depth": [3, 5, None],
-    "learning_rate": [0.1, 1],
+param_distributions = {
+    "n_estimators": randint(10, 200),
```
Comment (ogrisel) [inline python_scripts/ensemble_hyperparameters.py:119]:
```suggestion
    "n_estimators": [1, 2, 5, 10, 20, 50, 100, 200, 500],
```

File: python_scripts/ensemble_hyperparameters.py [hunk 6]
```diff
@@ -103,37 +112,46 @@
...
+    "max_leaf_nodes": randint(3, 300),
```
Comment (ogrisel) [inline python_scripts/ensemble_hyperparameters.py:120]:
```suggestion
    "max_leaf_nodes": [2, 5, 10, 20, 50, 100],
```

**Review by ogrisel — COMMENTED** [review]
I don't like `randint` because it would use a uniform prior over an integer range which is suboptimal. Let's rather use a fine-grained enough approximately exponential integer grid.

**Comment by ogrisel** [comment]
The output is not rendered anymore either...

File: python_scripts/ensemble_hyperparameters.py [hunk 7]
```diff
@@ -41,52 +45,57 @@
...
-# gradient-boosting that the algorithm fits the error of the previous tree in
-# the ensemble. Thus, fitting fully grown trees will be detrimental. Indeed, the
-# first tree of the ensemble would perfectly fit (overfit) the data and thus no
-# subsequent tree would be required, since there would be no residuals.
-# Therefore, the tree used in gradient-boosting should have a low depth,
-# typically between 3 to 8 levels. Having very weak learners at each step will
-# help reducing overfitting.
+# Let's first discuss the `max_depth` (or `max_leaf_nodes`) parameter. We saw
+# in the section on gradient-boosting that the algorithm fits the error of the
+# previous tree in the ensemble. Thus, fitting fully grown trees will be
```
Comment (ArturoAmorQ) [inline python_scripts/ensemble_hyperparameters.py:92]:
```suggestion
# previous tree in the ensemble. Thus, fitting fully grown trees would be
```

File: python_scripts/ensemble_hyperparameters.py [hunk 8]
```diff
@@ -103,37 +112,46 @@
...
 # %% [markdown]
+#
 # ```{caution}
 # Here, we tune the `n_estimators` but be aware that using early-stopping as
 # in the previous exercise will be better.
 # ```
+#
+# In this search, we see that the `learning_rate` is required to be large
+# enough, i.e. > 0.1. We also observe that for the best ranked models, having a
+# smaller `learning_rate`, will required more trees or a larger number of
```
Comment (ArturoAmorQ) [inline python_scripts/ensemble_hyperparameters.py:144]:
```suggestion
# smaller `learning_rate`, will require more trees or a larger number of
```

File: python_scripts/ensemble_hyperparameters.py [hunk 9]
```diff
@@ -103,37 +112,46 @@
...
+# leaves for each tree. However, this is particularly to diffcult to draw
```
Comment (ArturoAmorQ) [inline python_scripts/ensemble_hyperparameters.py:145]:
```suggestion
# leaves for each tree. However, it is particularly difficult to draw
```

**Comment by ogrisel** [comment]
I merged `main` to trigger a clean re-render without the CI cache problem.

**Review by ogrisel — APPROVED** [review]

**Pull request merged**